libcentrifugo/engineredis.go

```diff
@@ -427,11 +446,65 @@ func (e *RedisEngine) runPubSub() {
 	}
 }
 
-func (e *RedisEngine) publish(chID ChannelID, message []byte) (bool, error) {
+func (e *RedisEngine) publish(chID ChannelID, message []byte, opts *publishOpts) error {
 	conn := e.pool.Get()
 	defer conn.Close()
-	numSubscribers, err := redis.Int(conn.Do("PUBLISH", chID, message))
-	return numSubscribers > 0, err
+
+	var err error
+	if opts == nil {
+		// just publish message into channel.
+		_, err = e.pubScript.Do(conn, chID, message, "0")
+	} else {
+		// publish message into channel and add history message.
+		if opts.HistorySize > 0 && opts.HistoryLifetime > 0 {
+			messageJSON, err := json.Marshal(opts.Message)
```

Comment:
Huh so the history message is somehow different from the published payload? I guess I missed that. Is that necessary?